blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,701 @@
+        ObjectNode asJson = mapper.createObjectNode()
+                .put("id", order.getId())
+                .put("customerId", order.getCustomerId())
+                .put("orderDate", order.getOrderDate().toString());
+
+        ArrayNode items = asJson.putArray("lineItems");
+
+        for (OrderLine orderLine : order.getLineItems()) {
+        items.add(
+                mapper.createObjectNode()
+                .put("id", orderLine.getId())
+                .put("item", orderLine.getItem())
+                .put("quantity", orderLine.getQuantity())
+                .put("totalPrice", orderLine.getTotalPrice())
+                .put("status", orderLine.getStatus().name())
+            );
+        }
+
+        return new OrderCreatedEvent(order.getId(), asJson);
+    }
+
+    @Override
+    public String getAggregateId() {
+        return String.valueOf(id);
+    }
+
+    @Override
+    public String getAggregateType() {
+        return "Order";
+    }
+
+    @Override
+    public String getType() {
+        return "OrderCreated";
+    }
+
+    @Override
+    public JsonNode getPayload() {
+        return order;
+    }
+}
+----
+
+Note how https://github.com/FasterXML/jackson[Jackson's] `ObjectMapper` is used to create a JSON representation of the event's payload.
+
+Now let's take a look at the code that consumes any fired `ExportedEvent` and does the corresponding write to the outbox table:
+
+[source,java]
+----
+@ApplicationScoped
+public class EventSender {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    public void onExportedEvent(@Observes ExportedEvent event) {
+        OutboxEvent outboxEvent = new OutboxEvent(
+                event.getAggregateType(), event.getAggregateId(), event.getType(), event.getPayload()
+        );
+
+        entityManager.persist(outboxEvent);
+        entityManager.remove(outboxEvent);
+    }
+}
+----
+
+It's rather simple: for each event the CDI runtime will invoke the `onExportedEvent()` method.
+An instance of the `OutboxEvent` entity is persisted in the database -- and removed right away!
+
+This might be surprising at first.
+But it makes sense when remembering how log-based CDC works:
+it doesn't examine the actual contents of the table in the database, but instead it tails the append-only transaction log.
+The calls to `persist()` and `remove()` will create an `INSERT` and a `DELETE` entry in the log once the transaction commits.
+After that, Debezium will process these events (handling the `INSERT` and ignoring the `DELETE`).
+So we are able to capture the event added to the outbox table by means of CDC,
+but when looking at the contents of the table itself, it will always be empty.
+This means that no additional disk space is needed for the table
+(apart from the log file elements which will automatically be discarded at some point)
+and also no separate house-keeping process is required to stop it from growing indefinitely.
```

Comment:
Hmmm - I have very mixed feelings about this. This is related to the idea that events live in Kafka long term. Essentially it moves operational complexity from a well known industry space (relational DBs) towards a very poorly known industry space (kafka / k8s / ocp).

There are other complexities here too. I need to think about this.

Reply:
Per the previous comment thread - I'm OK with this approach, but we should add a line here that mentions that this approach is a corollary of the indefinite storage in Kafka (which again is one option of several). I'm just worried that I'll get customers telling me "This is the one and only way to implement outbox! Gunnar said so!"